snuba/subscriptions/scheduler.py

```diff
@@ -1,13 +1,259 @@
 import math
+from abc import ABC, abstractmethod
 from datetime import datetime, timedelta
-from typing import Iterator, List, Optional
+from enum import Enum
+from typing import (
+    Generic,
+    Iterator,
+    List,
+    Mapping,
+    MutableMapping,
+    Optional,
+    Sequence,
+    Tuple,
+    TypeVar,
+)
 
+from snuba import settings, state
 from snuba.subscriptions.data import PartitionId, Subscription, SubscriptionIdentifier
 from snuba.subscriptions.store import SubscriptionDataStore
 from snuba.utils.metrics import MetricsBackend
-from snuba.utils.scheduler import Scheduler, ScheduledTask
+from snuba.utils.scheduler import ScheduledTask, Scheduler
 from snuba.utils.types import Interval
 
+TSubscription = TypeVar("TSubscription")
+
+Tags = Mapping[str, str]
+
+
+class TaskBuilder(ABC, Generic[TSubscription]):
+    """
+    Takes a Subscription and a timestamp, decides whether we should
+    schedule that task at the current timestamp and provides the
+    task instance.
+    """
+
+    @abstractmethod
+    def get_task(
+        self, subscription: Subscription, timestamp: int
+    ) -> Optional[ScheduledTask[TSubscription]]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def reset_metrics(self) -> Sequence[Tuple[str, int, Tags]]:
+        raise NotImplementedError
+
+
+class ImmediateTaskBuilder(TaskBuilder[Subscription]):
+    """
+    Schedules a subscription as soon as possible
+    """
+
+    def __init__(self) -> None:
+        self.__count = 0
+
+    def get_task(
+        self, subscription: Subscription, timestamp: int
+    ) -> Optional[ScheduledTask[Subscription]]:
+        resolution = int(subscription.data.resolution.total_seconds())
+        if timestamp % resolution == 0:
+            self.__count += 1
+            return ScheduledTask(datetime.fromtimestamp(timestamp), subscription)
+        else:
+            return None
+
+    def reset_metrics(self) -> Sequence[Tuple[str, int, Tags]]:
+        metrics: Sequence[Tuple[str, int, Tags]] = [("tasks.built", self.__count, {})]
+        self.__count = 0
+        return metrics
+
+
+class JitteredTaskBuilder(TaskBuilder[Subscription]):
+    """
+    Schedules subscriptions applying a jitter to distribute subscriptions
+    evenly in the resolution period.
+
+    Each subscription is assigned a stable jitter which correspond to a
+    timestamp in the resolution interval.
+    For example if a subscription has a 60 seconds resolution, the jitter
+    is calculated from the subscription id (to stay constant) and it is
+    a number between 0 and 59.
+
+    That subscription is then scheduled when the timestamp % 60 is equal
+    to the jitter instead of being equal to 0.
+    This would spread the subscription evenly.
+
+    There is a setting to define the maximum resolution the jitter applies.
+
+    The time range of the query does not include the jitter. So if the
+    jitter for a 60 seconds resolution query is 4 seconds. The query is
```

Comment:
If we are guaranteeing alerts be accurate to the previous minute, then I think a jitter of 59 seconds would be too much. The alerts would be essentially 1 minute delayed.

Reply:
This is true, alerts can be delayed by an average that is half the resolution. Though the 1 minute delay already exists (there is an artificial 1 minute delay in the time range of the query to give time to as many events as possible to go through the ingestion pipeline. The delay we are introducing here is smaller than that.

Reply:
If there is already a 1 minute delay, then this is adding in the worst case effectively an entire extra minute. If you're saying that at least 1/60th of subscriptions having a 2 minute delay is not going to bother clients, that's fine. I just thought maybe we should change the jitter to be at most 30 seconds or something like that.

Reply:
The issue with 30 seconds is that with the current scheduling we are already scheduling across the first 25 seconds of each minute (because of batch flushing time). I would start by spreading evenly, then we can experiment ways to tune the jitter down to find the balance where clickhouse is fine, we do not backlog and we minimize the delay.